llvm/lib/Target/DirectX/DXILRootSignature.h

```diff
@@ -0,0 +1,91 @@
+//===- DXILRootSignature.h - DXIL Root Signature helper objects
+//---------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file This file contains helper objects and APIs for working with DXIL
+///       Root Signatures.
+///
+//===----------------------------------------------------------------------===//
+
+#include "llvm/IR/DiagnosticInfo.h"
+#include "llvm/IR/Metadata.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+#include <optional>
+
+namespace llvm {
+namespace dxil {
+
+enum class RootSignatureElementKind {
+  None = 0,
+  RootFlags = 1,
+  RootConstants = 2,
+  RootDescriptor = 3,
+  DescriptorTable = 4,
+  StaticSampler = 5
+};
+
+struct ModuleRootSignature {
+  uint32_t Flags = 0;
+  ModuleRootSignature() { Ctx = nullptr; };
+  static std::optional<ModuleRootSignature> analyzeModule(Module &M,
+                                                          const Function *F);
+
+private:
+  LLVMContext *Ctx;
+
+  ModuleRootSignature(LLVMContext *Ctx) : Ctx(Ctx) {}
+
+  bool parse(NamedMDNode *Root, const Function *F);
+  bool parseRootSignatureElement(MDNode *Element);
+  bool parseRootFlags(MDNode *RootFlagNode);
+
+  bool validate();
+
+  bool reportError(Twine Message, DiagnosticSeverity Severity = DS_Error);
+};
+
+using OptionalRootSignature = std::optional<ModuleRootSignature>;
```

Comment:
```suggestion
```

Does this really help with readability at all?  For example, to figure out what `RootSignatureAnalysis::Result` is I have to look up the definition of `Result` and then I have to look up the definition of `OptionalRootSignature` and then I need to look up `ModuleRootSignature`.

The inconsistency between whether or not it is "RootSignature" or "ModuleRootSignature" is confusing as well.